### PR TITLE
logging: Generalize ring buffer in memory backend

### PR DESCRIPTION
**Pull request by finikorg**
Make ring buffer in memory logging back-end available for other platforms.

### CODE REVIEW

File: subsys/logging/Kconfig [hunk 1]
```diff
@@ -497,26 +497,25 @@ config LOG_BACKEND_NET_SYST_ENABLE
 
 endif # LOG_BACKEND_NET
 
-config LOG_BACKEND_ADSP
-	bool "Enable Intel ADSP backend"
-	depends on SOC_INTEL_APL_ADSP
+config LOG_BACKEND_RB
+	bool "Enable mewmory ring buffer backend"
```
Comment (dcpleung) [inline subsys/logging/Kconfig]: s/mewmory/memory/

**Review by dcpleung — COMMENTED** [review]
Just a typo.

**Review by dcpleung — APPROVED** [review]

File: subsys/logging/Kconfig [hunk 2]
```diff
@@ -497,26 +497,32 @@ config LOG_BACKEND_NET_SYST_ENABLE
...
 	  Address of the ring buffer in the memory.
 
-config LOG_BACKEND_ADSP_RINGBUF_SIZE
+config LOG_BACKEND_RB_MEM_SIZE
 	hex "Size of the ring buffer"
 	help
 	  Size of the ring buffer.
 
-endif # LOG_BACKEND_ADSP
+config LOG_BACKEND_RB_BUF_SIZE
```
Comment (nordic-krch) [inline subsys/logging/Kconfig]: why not `LOG_BACKEND_RB_MSG_SIZE` it is a bit confusing because it suggests some buffer dedicated for ring buffer and description says "size of message slot".

File: subsys/logging/Kconfig [hunk 3]
```diff
@@ -497,26 +497,32 @@ config LOG_BACKEND_NET_SYST_ENABLE
...
+config LOG_BACKEND_RB_MEM_BASE
+	hex "Ring buffer memory base address"
 	help
 	  Address of the ring buffer in the memory.
 
-config LOG_BACKEND_ADSP_RINGBUF_SIZE
+config LOG_BACKEND_RB_MEM_SIZE
 	hex "Size of the ring buffer"
 	help
 	  Size of the ring buffer.
```
Comment (nordic-krch) [inline subsys/logging/Kconfig]: must be multiply of LOG_BACKEND_RB_BUF_SIZE

File: subsys/logging/Kconfig (hunk 2 above)
Comment (finikorg) [reply subsys/logging/Kconfig]: Then `MSG` suggest that this is message size, while it is not.

Comment (nordic-krch) [reply subsys/logging/Kconfig]: LOG_BACKEND_RB_SLOT_SIZE ?
LOG_BACKEND_RB_MSG_SLOT_SIZE ?

Comment (finikorg) [reply subsys/logging/Kconfig]: I will use something with `SLOT`.

Comment (nordic-krch) [reply subsys/logging/Kconfig]: currently name `LOG_BACKEND_RB_BUF_SIZE` does not correspond to the description; " Size of the message slot inside ring buffer"

File: subsys/logging/Kconfig (hunk 1 above)
Comment (finikorg) [reply subsys/logging/Kconfig]: Fixed

**Review by nordic-krch — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by tyrantlizard77 — APPROVED** [review]

File: subsys/logging/Kconfig [hunk 4]
```diff
@@ -497,26 +497,33 @@ config LOG_BACKEND_NET_SYST_ENABLE
...
 	hex "Size of the ring buffer"
 	help
-	  Size of the ring buffer.
+	  Size of the ring buffer. Must be multiply of the message slot
+	  size.
 
-endif # LOG_BACKEND_ADSP
+config LOG_BACKEND_RB_SLOT_SIZE
+	hex "Size of the message slot"
+	default 0x40
```
Comment (jhedberg) [inline subsys/logging/Kconfig]: What's the reason you use `hex` and not `int`? Wouldn't decimal notation be more natural?

File: subsys/logging/Kconfig [hunk 5]
```diff
@@ -497,26 +497,33 @@ config LOG_BACKEND_NET_SYST_ENABLE
...
 	help
 	  Address of the ring buffer in the memory.
 
-config LOG_BACKEND_ADSP_RINGBUF_SIZE
+config LOG_BACKEND_RB_MEM_SIZE
 	hex "Size of the ring buffer"
 	help
-	  Size of the ring buffer.
+	  Size of the ring buffer. Must be multiply of the message slot
+	  size.
```
Comment (jhedberg) [inline subsys/logging/Kconfig]: Unrelated to what this PR is trying to do, but would it make sense to rethink this config option and instead call it `LOG_BACKEND_RB_SLOT_COUNT`. That way you don't need to remind the user of using multiples of `LOG_BACKEND_RB_SLOT_SIZE`.

**Review by nashif — APPROVED** [review]

File: subsys/logging/Kconfig (hunk 5 above)
Comment (finikorg) [reply subsys/logging/Kconfig]: `SLOT_COUNT` still need to be multiples of memory size so nothing changes.
There are advantages using `SLOT_SIZE` like it is human understandable size of the memory slot which we can adjust if we see slots are used inefficiently.

File: subsys/logging/Kconfig (hunk 4 above)
Comment (finikorg) [reply subsys/logging/Kconfig]: I can change to decimal here

File: subsys/logging/Kconfig (hunk 5 above)
Comment (jhedberg) [reply subsys/logging/Kconfig]: > SLOT_COUNT still need to be multiples of memory size so nothing changes.

Why? It's not a memory size as-is. You'd get the total memory size by multiplying it with the SLOT_SIZE.

> There are advantages using SLOT_SIZE like it is human understandable size of the memory slot which we can adjust if we see slots are used inefficiently.

That's fair. Though since you use the word "Must" I guess you should back that up by adding a BUILD_ASSERT? Alternatively you could reword this as "*should* be a multiple, or memory will be wasted"

Comment (finikorg) [reply subsys/logging/Kconfig]: We have `BUILD_ASSERT` here: https://github.com/finikorg/zephyr/blob/41f6adb1b56dd118e49f3f43d520483df6c65297/subsys/logging/log_backend_rb.c#L28

>Why? It's not a memory size as-is. You'd get the total memory size by multiplying it with the SLOT_SIZE.

We usually get total available memory window (like https://github.com/finikorg/zephyr/blob/41f6adb1b56dd118e49f3f43d520483df6c65297/soc/xtensa/intel_apl_adsp/Kconfig.defconfig#L55) If we have count * slot size we need to check that it fits to the window.

Comment (jhedberg) [reply subsys/logging/Kconfig]: @finikorg thanks for clarifying!

File: subsys/logging/Kconfig [hunk 6]
```diff
@@ -497,26 +497,33 @@ config LOG_BACKEND_NET_SYST_ENABLE
...
-config LOG_BACKEND_ADSP_RINGBUF_BASE
-	hex "Ringbuf base address"
+config LOG_BACKEND_RB_MEM_BASE
+	hex "Ring buffer memory base address"
 	help
 	  Address of the ring buffer in the memory.
 
-config LOG_BACKEND_ADSP_RINGBUF_SIZE
+config LOG_BACKEND_RB_MEM_SIZE
 	hex "Size of the ring buffer"
```
Comment (jhedberg) [inline subsys/logging/Kconfig]: I guess this one should also be made `int` for consistency if you change the SLOT_SIZE to be `int`?

Comment (finikorg) [reply subsys/logging/Kconfig]: Changed to `int`.

**Comment by jhedberg** [comment]
@finikorg @dcpleung @nashif do we want to target this for 2.2? It's not strictly a bug fix, so may need TSC approval, however since these Kconfig variables haven't been part of any release yet now would be the least painful time to fix this.

**Comment by dcpleung** [comment]
I don't think there is a need to get this in 2.2, right? It looks more like a new feature.

**Comment by jhedberg** [comment]
> I don't think there is a need to get this in 2.2, right? It looks more like a new feature.

Yes, it's basically new feature. It's also kind of an API break in that any out-of-tree app referencing the old Kconfig defines would fail to build. If we never release the old Kconfig names this issue would never happen, which is why I raised the question. Anyway, I'm fine if the consensus is to postpone this to 2.3.

**Comment by andrewboie** [comment]
I think this should wait until 2.3.

**Comment by maksimmasalski** [comment]
@finikorg What tests are using to test that feature? Does it necessary to use special hardware or only qemu?

**Comment by finikorg** [comment]
> @finikorg What tests are using to test that feature? Does it necessary to use special hardware or only qemu?

There is not tests for this backend. It can be used on any hardware, if you allocate shared memory area in qemu, it should be possible to use it also there. I have tested it on `up_squared` board with instructions here: https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/xtensa/up_squared_adsp/doc/index.rst#debugging

**Comment by maksimmasalski** [comment]
> > @finikorg What tests are using to test that feature? Does it necessary to use special hardware or only qemu?
> 
> There is not tests for this backend. It can be used on any hardware, if you allocate shared memory area in qemu, it should be possible to use it also there. I have tested it on `up_squared` board with instructions here: https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/xtensa/up_squared_adsp/doc/index.rst#debugging

Can I use reel board instead?

**Comment by finikorg** [comment]
> Can I use reel board instead?

You can write to ring buffer in memory in almost every board, so it can be tested there also. But you need to be able to read what you have written.

I recommend you to use `up_squared`, I think you have one.

**Comment by maksimmasalski** [comment]
> > Can I use reel board instead?
> 
> You can write to ring buffer in memory in almost every board, so it can be tested there also. But you need to be able to read what you have written.
> 
> I recommend you to use `up_squared`, I think you have one.

I have, but in Shanghai office, now I'm WFH due to the coronavirus outbreak, my task is to validate that PR. At home I have only that UP squared board https://devices.amazonaws.com/detail/a3G0L00000AANtUUAX/UP-Squared-AI-Vision-Developer-Kit

**Comment by maksimmasalski** [comment]
> > @finikorg What tests are using to test that feature? Does it necessary to use special hardware or only qemu?
> 
> There is not tests for this backend. It can be used on any hardware, if you allocate shared memory area in qemu, it should be possible to use it also there. I have tested it on `up_squared` board with instructions here: https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/xtensa/up_squared_adsp/doc/index.rst#debugging

Do you have any advice how to test it in qemu?

**Comment by finikorg** [comment]
> @finikorg Do you have any advice how to test it in qemu?

I have never tried myself but if Qemu can share memory between host and guest then you can specify the guest memory address in Zephyr configuration, on the host (Linux) it is probably some device you can access with our logging frontend tools.
